solr/core/src/java/org/apache/solr/core/CoreContainer.java

```diff
@@ -2264,13 +2264,13 @@ public void run() {
           // any cores to close.
         }
       }
-      for (SolrCore removeMe = solrCores.getCoreToClose();
-           removeMe != null && !container.isShutDown();
```

Comment:
There was a bug here: when the CoreContainer was shutting down it was possible to get the core to close (and remove it from the pendingClose list) and then simply skip the core by exiting this method without closing it. It resulted in unreleased Core as well as other related objects (Directory, etc).